New Post: Debezium Server streaming from MySQL performance (#1213)
debezium/debezium.github.io: _posts/2026-02-02-measuring-debezium-server-performance-mysql-streaming.adoc

---
layout: post
title:  "Measuring Debezium Server performance when streaming MySQL to Kafka"
date:   2026-01-08
tags:   [debezium-server, performance, streaming, mysql, docker]
author: alvarvg
---

Performance is a critical concern when implementing Change Data Capture (CDC) solutions in production environments.
In this post, I am going to share how we have measured Debezium Server performance (deployed in a Docker container)
while streaming changes from MySQL to Kafka. All of this with the default configurations for Debezium, MySQL and Kafka.

+++<!-- more -->+++

This post does not aim to compare Debezium Server with Kafka Connect nor to identify absolute throughput limits. 
Instead, it focuses on understanding how Debezium Server behaves under realistic workloads using default configurations, 

> and whether it can sustain real-time change data capture in lightweight architectures.

Reviewer: I would also like to underline that the system under CDC (the database) is not stressed in a real-world scenario (such as long-running transactions), which is not taken into account in our test case due to its complexity.

Author: Thanks! I totally forgot about this! I will add it and work on this tomorrow!

Author: Applied thanks!

Now that we are familiar with all the components, let us dive into the technical details. 

Starting with the hardware, we have deployed two EC2 machines in AWS of type "c5a.xlarge". Each instance provides 4 vCPUs and 8 GiB of memory, ensuring sufficient resources for both workload generation and CDC processing.
We have called them Observed and Observer (in order to separate the monitoring load from the machine where Debezium and MySQL are running). In the Observer instance, we have deployed Prometheus, Grafana and YCSB
(the software piece generating the load). In the Observed instance, we have a single Kafka container (although it is not really necessary and another software can be used),
Debezium Server and MySQL. 

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/benchmark_architecture_blueprint.png" style="max-width:70%;" class="responsive-image" alt="Benchmarking architecture">
++++
*Benchmarking architecture*
====

As described above, we are using some software pieces, most of them are already known, except for YCSB. Yahoo! Cloud Serving Benchmark (YCSB), is a tool created by the Yahoo

> team and released into open source as of 04/23/2010, developed to evaluate the performance of different cloud serving stores. This is the link to their GitHub project: https://ycsb.site[https://ycsb.site].

Reviewer 2: ```suggestion
team and released into open source as of 04/23/2010, is used to evaluate the performance of different cloud serving stores. You can access YCSB via its GitHub project: https://ycsb.site[https://ycsb.site].
```

If you want to reproduce this scenario, it is quite simple. Go to debezium-performance[ADD_REPO_URL_HERE] repo, clone it and cd into the infrastructure-automation folder and then run terraform apply. 

> This will automatically create the EC2 instances, provision them with ansible (copy files and install docker) and start the scenario.

Reviewer 2: ```suggestion
Running the command automatically creates the EC2 instances, provisions them with Ansible (copying files and installing docker), and start the scenario.
```

> Logging in Grafana will allow you to see how everything evolves and progresses (inside Grafana you will need to navigate to the MySQL Streaming dashboard).

Reviewer 2: ```suggestion
Log in to Grafana to see how everything evolves and progresses. Within Grafana you must navigate to the MySQL Streaming dashboard.
```

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/benchmark_completed_scenario.png" style="max-width:70%;" class="responsive-image" alt="Streaming dashboard from a completed scenario">
++++
*Streaming dashboard from a completed scenario*
====

> With the technology stack clearly defined and reproducible, we can now focus on how performance will be measured and which metrics will be considered:

Reviewer 2: ```suggestion
With the technology stack clearly defined and reproducible, we can now focus on how to measure  performance and which of the following metrics to consider:
```

> - CPU and RAM Percentage usage: We are monitoring the resource usage from the Debezium Server container.
> - Network Input and Output: We are also taking into account the communications performed by the container.
> - Disk Reads and Writes: To check how much disk space/bandwidth is needed.
> - Throughput: This is the main value we are going to focus on (based on this value, we can estimate the operation latency) and we are comparing this to MySQL throughput. 

Reviewer 2: ```suggestion
CPU and RAM Percentage usage:: Monitors resource usage within the Debezium Server container.
Network Input and Output:: Monitors the communications performed by the container.
Disk Reads and Writes:: Reports the amount of disk space and bandwidth that is consumed.
Throughput:: Reports the amount of data that Debezium processes. By comparing this value to the MySQL throughput, we can estimate the operational latency. This is the main value to focus on.
```
As in the documentation, I'd recommend that you format this list as a description list vs. an unordered list. The description list formatting provides greater readability/scannability by highlighting (via bold text) each term in the list, and indenting the descriptions.

For all the given metrics, we are calculating the average over the duration of the test. This way, we provide a stable value, and any possible spikes that may exist do not affect

> the final result and are included in it. These are a few more detailed panels of the dashboard.

Reviewer 2: ```suggestion
the final result and are included in it. 
The dashboard provides several more useful details.
```

[.centered-image.responsive-image]
====
++++
<img src="/assets/images/2026-01-08-measuring-debezium-server-performance-mysql-streaming/cpu_memory_consumption.png" style="max-width:70%;" class="responsive-image" alt="CPU and Memory consumption during one execution">
++++
*CPU and Memory consumption during one execution*
====

> In the picture above, you can see a few of the panels related to the resource consumption. The two on top are related to the CPU usage, which is the calculated percentage of usage for a 4 vCPUs instance.

Reviewer 2: ```suggestion
The preceding figure shows panels that provide information about resource consumption. The first two **Container Memory Usage** panels report the memory consumption, in MB through the duration of the test. The first two of these panels represent the evolution over time, and the panels that follow present the all-time average. 
```
Looking at the image, `cpu_memory-consumptio.png`, I think that your description is reversed?

> And the other ones are the memory consumption in MB throughout the execution. The two on the left represent the evolution over time, and the ones on the right are the all-time average. 

Reviewer 2: ```suggestion
The remaining panels report the container CPU usage, which is the calculated percentage of usage for a 4 vCPUs instance.
```

Reviewer 2: In addition to reversing the order of the descriptions to match the PNG file, I'd suggest that you remove directional language (_above_, _top_, _left_, _right_), in accordance with web content accessibility guidelines (WCAG). See https://stephaniefox.ca/ux-blog/directional-language/

With these results and conclusions in mind, if I had to make a recommendation to someone who wants to implement CDC in their pipeline, I would say the following:

1. *Start with Debezium Server and default configurations.* For small-to-medium projects, Debezium Server provides an excellent starting point without the operational complexity of Kafka Connect or Kubernetes.
The results show that default configurations already deliver stable, predictable and near real-time CDC performance, allowing teams to validate their use cases before investing time in advanced tuning or more complex architectures.
2. *Design tables and schemas with CDC in mind.* Performance is strongly influenced by table-level characteristics rather than raw operations counts. Careful schema design can have a significant impact on CDC throughput and latency.
Optimizing table structures will often yield greater benefits than tuning Debezium itself.

> 3. *Treat the database as the primary throughput limiter.* The benchmark shows that Debezium Server can keep up with MySQL in near real time and that overall throughput is bounded by database's write capacity.

Reviewer 3: How about a follow-up - trying to beef up the database to saturate Debezium and identify the breaking point?
